packages/grid/x-data-grid/src/components/menu/columnMenu/GridColumnMenu.tsx

```diff
@@ -1,47 +1,147 @@
 import * as React from 'react';
 import PropTypes from 'prop-types';
-import { GridColumnMenuContainer } from './GridColumnMenuContainer';
+import { GridColumnMenuLookup, GridColumnMenuValue } from '../../../hooks/features/columnMenu';
 import { GridColumnMenuProps } from './GridColumnMenuProps';
-import { GridColumnsMenuItem } from './GridColumnsMenuItem';
-import { GridFilterMenuItem } from './GridFilterMenuItem';
-import { HideGridColMenuItem } from './HideGridColMenuItem';
-import { SortGridMenuItems } from './SortGridMenuItems';
 import { useGridApiContext } from '../../../hooks/utils/useGridApiContext';
 
-const GridColumnMenu = React.forwardRef<HTMLUListElement, GridColumnMenuProps>(
-  function GridColumnMenu(props: GridColumnMenuProps, ref) {
-    const { hideMenu, currentColumn } = props;
-    const apiRef = useGridApiContext();
-
-    const defaultButtons = [
-      <SortGridMenuItems onClick={hideMenu} column={currentColumn!} />, // TODO update types to allow `onClick` and `column` to be optional
-      <GridFilterMenuItem onClick={hideMenu} column={currentColumn!} />,
-      <HideGridColMenuItem onClick={hideMenu} column={currentColumn!} />,
-      <GridColumnsMenuItem onClick={hideMenu} column={currentColumn!} />,
-    ];
-
-    const preProcessedButtons = apiRef.current.unstable_applyPipeProcessors(
-      'columnMenu',
-      defaultButtons,
-      currentColumn,
-    );
+const GridColumnMenu = (props: GridColumnMenuProps) => {
+  const {
+    hideMenu,
+    currentColumn,
+    defaultMenuItems,
+    defaultVisibleSlots,
+    columnMenuItems: userMenuItems,
+  } = props;
+
+  const apiRef = useGridApiContext();
 
-    return (
-      <GridColumnMenuContainer ref={ref} {...props}>
-        {preProcessedButtons.map((button: any, index: number) =>
-          React.cloneElement(button, { key: index, onClick: hideMenu, column: currentColumn }),
-        )}
-      </GridColumnMenuContainer>
+  const preProcessedValue = apiRef.current.unstable_applyPipeProcessors(
+    'columnMenu',
+    { items: defaultMenuItems, visibleSlots: defaultVisibleSlots },
+    currentColumn,
+  );
+
+  const extendedColumnMenuItems = React.useMemo(() => {
+    if (!userMenuItems || !Object.keys(userMenuItems).length) {
+      return preProcessedValue.items;
+    }
+    // Overrides for default items
+    const userSlots = new Set<GridColumnMenuLookup['slot']>(
+      Object.keys(userMenuItems) as Array<GridColumnMenuLookup['slot']>,
     );
-  },
-);
+
+    const overridenItems = preProcessedValue.items.reduce((acc, item) => {
```

Comment:
Would it make sense to make `preProcessedValue.items` a `Record` instead of `Array` and use `preProcessedValue.visibleSlots` for ordering?
Pros:
- easier to merge with `userMenuItems`
- easier to transform `filteredSlots` into `filteredColumnMenuItems` (no need to use `.find` inside `.reduce`. Maybe even get rid of `filteredColumnMenuItems` and just use `filteredSlots` (we can easily access data from `extendedColumnMenuItems`)

Reply:
Exactly, I was thinking around the same lines, thanks for the feedback